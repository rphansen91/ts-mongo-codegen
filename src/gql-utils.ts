import {
  ArgumentNode,
  buildSchema,
  DirectiveNode,
  GraphQLInputObjectType,
  GraphQLInt,
  GraphQLList,
  GraphQLNamedType,
  GraphQLObjectType,
  GraphQLSchema,
  StringValueNode,
} from 'graphql'
import capitalize from 'lodash/capitalize'
import get from 'lodash/fp/get'
import last from 'lodash/last'
import mapValues from 'lodash/mapValues'

export type AugmentConfig = {
  name?: string
}

function hasCollectionComment(str: string) {
  const match = str.match(/@collection\(name: ?"(\w+)"\) (@crud)/)
  if (!match) return null
  return {
    name: match[1],
    crud: Boolean(match[2]),
  }
}

export function findDirective(name: string, type: GraphQLNamedType) {
  return type?.astNode?.directives?.find?.((d) => d.name.value === name)
}

export function findDirectiveOption(name: string, directive: DirectiveNode | undefined) {
  return directive?.arguments?.find?.((d) => d.name.value === name)
}
export function findArgumentValue(argument: ArgumentNode | undefined) {
  return (argument?.value as StringValueNode)?.value
}

export type CollectionMap = ReturnType<typeof directiveTypeMap>
export function directiveTypeMap(schema: GraphQLSchema, name: string) {
  const typesMap = schema.getTypeMap()
  return Object.keys(typesMap).reduce((acc, typeName: string) => {
    const type = typesMap[typeName]
    const collection = findDirective(name, type)
    if (!collection) return acc
    acc[typeName] = type as GraphQLObjectType
    return acc
  }, {} as { [key: string]: GraphQLObjectType })
}

export function isDirectiveNode(v: any): v is DirectiveNode {
  return Boolean(v)
}

export function pluralize(str: string) {
  if (last(str) === 's') return str
  return `${str}s`
}

export type MongoTypeMap = ReturnType<typeof buildMongoTypeMap>
export function buildMongoTypeMap(schema: GraphQLSchema, config?: AugmentConfig) {
  const collectionMap = directiveTypeMap(schema, 'collection')
  const [pageSchemaMap, pageTypeMap] = buildPageMap({ collectionMap })
  const [filterSchemaMap, filterTypeMap] = buildFieldMap('filter', { collectionMap })
  const [insertSchemaMap, insertTypeMap] = buildFieldMap('insert', { collectionMap })
  const [unsetSchemaMap, unsetTypeMap] = buildFieldMap('unset', { collectionMap })
  const [setSchemaMap, setTypeMap] = buildFieldMap('set', { collectionMap })
  const [incSchemaMap, incTypeMap] = buildFieldMap('inc', { collectionMap })
  const [decSchemaMap, decTypeMap] = buildFieldMap('dec', { collectionMap })
  return {
    collectionMap,
    pageSchemaMap,
    filterSchemaMap,
    insertSchemaMap,
    unsetSchemaMap,
    setSchemaMap,
    incSchemaMap,
    decSchemaMap,
    pageTypeMap,
    filterTypeMap,
    insertTypeMap,
    unsetTypeMap,
    setTypeMap,
    incTypeMap,
    decTypeMap,
  }
}

export type CollectionMongoTypes = ReturnType<typeof getCollectionMongoTypes>
export function getCollectionMongoTypes(mongoTypeMap: MongoTypeMap, type: string) {
  return {
    collection: mongoTypeMap.collectionMap[type],
    pageSchema: mongoTypeMap.pageSchemaMap[type],
    filterSchema: mongoTypeMap.filterSchemaMap[type],
    insertSchema: mongoTypeMap.insertSchemaMap[type],
    unsetSchema: mongoTypeMap.unsetSchemaMap[type],
    setSchema: mongoTypeMap.setSchemaMap[type],
    incSchema: mongoTypeMap.incSchemaMap[type],
    decSchema: mongoTypeMap.decSchemaMap[type],
    pageType: mongoTypeMap.pageTypeMap[type],
    filterType: mongoTypeMap.filterTypeMap[type],
    insertType: mongoTypeMap.insertTypeMap[type],
    unsetType: mongoTypeMap.unsetTypeMap[type],
    setType: mongoTypeMap.setTypeMap[type],
    incType: mongoTypeMap.incTypeMap[type],
    decType: mongoTypeMap.decTypeMap[type],
  }
}

export function buildPageMap({ collectionMap }: { collectionMap: CollectionMap }) {
  const types = mapValues(collectionMap, (collection, name) => {
    return new GraphQLObjectType({
      name: `${name}Page`,
      fields: {
        total: {
          type: GraphQLInt,
        },
        data: {
          type: new GraphQLList(collection),
        },
      },
    })
  })
  const schemas = mapValues(types, (type, name) => {
    return new GraphQLSchema({
      types: [type],
      query: new GraphQLObjectType({ name: 'Query', fields: {} }),
    })
  })
  return tuple(schemas, types)
}

export function buildFieldMap(type: string, { collectionMap }: { collectionMap: CollectionMap }) {
  const schemas = mapValues(collectionMap, (collection, name) => {
    const fields = collection.astNode?.fields?.filter((field) => {
      return field.directives?.find((d) => d.name.value === type)
    })
    const types = fields
      ?.map((field) => `${field.name.value}: ${get('type.name.value', field)}`)
      .join('\n')
    if (!types?.length) return null
    return buildSchema(`input ${name}${capitalize(type)} {
      ${types}
    }`)
  })
  const types = mapValues(schemas, (schema, name) => {
    if (!schema) return null
    return schema.getType(`${name}${capitalize(type)}`) as GraphQLInputObjectType
  })
  return tuple(schemas, types)
}

export function tuple<A, B>(a: A, b: B) {
  return [a, b] as [A, B]
}

export function gqlTypeToTypescript(type: GraphQLInputObjectType | null) {
  const fields = type?.getFields?.() ?? {}
  return `{ ${Object.keys(fields)
    .map((k) => `${k}?: ${gqlScalarToTypescript(fields[k].type.toString())}`)
    .join(', ')} }`
}

export function gqlTypeToTypescriptUnset(type: GraphQLInputObjectType | null) {
  const fields = type?.getFields?.() ?? {}
  return `{ ${Object.keys(fields)
    .map((k) => `${k}?: 1`)
    .join(', ')} }`
}

function gqlScalarToTypescript(type: string) {
  switch (type) {
    case 'String':
      return 'string'
    case 'Float':
    case 'Int':
      return 'number'
    case 'Boolean':
      return 'boolean'
    default:
      return ''
  }
}
