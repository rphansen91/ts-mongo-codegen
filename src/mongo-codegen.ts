import { mongoTypeDefs } from './mongo-types'
import { PluginFunction } from '@graphql-codegen/plugin-helpers'
import { StringValueNode, DirectiveNode, GraphQLNamedType } from 'graphql'
import capitalize from 'lodash/capitalize'
import camelCase from 'lodash/camelCase'

export const addToSchema = mongoTypeDefs

export type Config = {
  name: 'string'
}

export const plugin: PluginFunction<Partial<Config>> = (schema, documents, config, info) => {
  const typesMap = schema.getTypeMap()
  const collections = Object.keys(typesMap)
    .map((typeName: string) => {
      const type = typesMap[typeName]
      const collection = findDirective('collection', type)
      if (!collection) return null
      return buildMongoCollection(typeName, collection)
    })
    .filter(isGeneratedCollection)

  const content = `import { Db, Collection } from 'mongodb'

${collections.map(({ exports }) => exports.map(generateExport).join('\n')).join('\n\n')}

export function mongoCollectionFactory (db: Db) {
  ${collections.map(({ factories }) => factories.map(generateExecute)).join('\n')}

  return {
    ${collections.map(({ factories }) => factories.map(({ name }) => name).join(',\n')).join(',\n')}
  }
}`

  return content
}

function findDirective(name: string, type: GraphQLNamedType) {
  return type?.astNode?.directives?.find?.(d => d.name.value === name)
}

type GeneratedCollection = ReturnType<typeof buildMongoCollection>
function buildMongoCollection(typeName: string, node: DirectiveNode) {
  const nameArgument = node?.arguments?.find?.(d => d.name.value === 'name')
  const collectionName = (nameArgument?.value as StringValueNode)?.value
  const name = camelCase(collectionName)
  const capitalName = capitalize(name)
  const collectionType = generate(`I${capitalName}Collection`, `Collection<${typeName}>`, 'type')
  const getCollectionType = generate(
    `get${capitalName}Collection`,
    `(db: Db) => db.collection<${typeName}>('${collectionName}')`,
    'const'
  )
  return {
    name,
    typeName,
    capitalName,
    collectionName,
    getCollectionType,
    exports: [collectionType, getCollectionType],
    factories: [
      {
        name,
        value: getCollectionType.name,
        args: ['db']
      }
    ]
  }
}

function isGeneratedCollection(v: any): v is GeneratedCollection {
  return v?.name
}

type Generated = ReturnType<typeof generate>
function generate(name: string, value: string, type = 'const') {
  return { name, value, type }
}

function generateDeclaration({ name, value, type }: Generated) {
  return `${type} ${name} = ${value}`
}

function generateExport({ name, value, type }: Generated) {
  return `export ${type} ${name} = ${value}`
}

function generateExecute(factory: GeneratedCollection['factories'][0]) {
  return generateDeclaration({
    type: 'const',
    name: factory.name,
    value: `${factory.value}(${(factory.args ?? []).join(', ')})`
  })
}
