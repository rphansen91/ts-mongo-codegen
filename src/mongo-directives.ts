import { ObjectID } from 'mongodb'
import { GraphQLScalarType, Kind } from 'graphql'

export const graphqlTypeObjectId = new GraphQLScalarType({
  name: 'ObjectId',
  description: 'Mongo Object Id',
  parseValue(value) {
    return new ObjectID(value)
  },
  serialize(value) {
    return value.toString() // value sent to the client
  },
  parseLiteral(ast) {
    if (ast.kind === Kind.STRING) {
      return new ObjectID(ast.value)
    }
    return null
  }
})
