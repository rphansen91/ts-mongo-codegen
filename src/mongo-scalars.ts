import { ObjectID } from 'mongodb'
import { GraphQLScalarType, Kind } from 'graphql'

export const graphqlTypeObjectId = new GraphQLScalarType({
  name: 'ObjectId',
  description: 'Mongo Object Id',
  parseValue(value) {
    return new ObjectID(value)
  },
  serialize(value) {
    return value.toString() // value sent to the client
  },
  parseLiteral(ast) {
    if (ast.kind === Kind.STRING) {
      return new ObjectID(ast.value)
    }
    return null
  }
})

export const graphqlTypeDate = new GraphQLScalarType({
  name: 'Date',
  description: 'Date custom scalar type',
  parseValue(value) {
    if (!value) return null
    return new Date(value) // value from the client
  },
  serialize(value) {
    if (value && value.toISOString) return value.toISOString()
    if (value && value.toString) return value.toString()
    return null // value sent to the client
  },
  parseLiteral(ast) {
    // if (ast.kind === Kind.INT) {
    //   return new Date(ast.value) // ast value is always in string format
    // }
    if (ast.kind === Kind.STRING) {
      return new Date(ast.value) // ast value is always in string format
    }
    return null
  }
})
