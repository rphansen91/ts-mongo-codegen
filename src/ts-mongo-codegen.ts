export * from './mongo'
export * from './mongo-types'
export * from './mongo-codegen'
export * from './mongo-augment'
export * from './mongo-scalars'
