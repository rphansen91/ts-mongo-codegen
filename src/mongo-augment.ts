import {
  GraphQLSchema,
  Kind,
  print,
  DefinitionNode,
  GraphQLObjectType,
  DirectiveNode,
  GraphQLString,
  GraphQLList,
  GraphQLInputObjectType,
  GraphQLNonNull,
  buildSchema,
  buildASTSchema,
  GraphQLNamedType,
  printSchema,
  GraphQLInt,
  isSchema,
} from 'graphql'
import { makeExecutableSchema, IExecutableSchemaDefinition } from '@graphql-tools/schema'
import { mapSchema, MapperKind, ObjectTypeMapper } from '@graphql-tools/utils'
import { mergeType, mergeGraphQLTypes, mergeSchemas } from '@graphql-tools/merge'
import { directiveTypeMap } from './gql-utils'
import values from 'lodash/values'
import capitalize from 'lodash/capitalize'
import merge from 'lodash/merge'
import last from 'lodash/last'
import mapValues from 'lodash/mapValues'
import first from 'lodash/first'
import get from 'lodash/fp/get'
import { graphqlTypeObjectId } from './mongo-scalars'
import { camelCase, isNull, negate } from 'lodash'
import gql from 'graphql-tag'

type AugmentConfig = {
  name?: string
}

export const augmentResolvers = ({
  resolvers,
  config,
}: {
  resolvers: any
  config: AugmentConfig
}) => {
  return resolvers
}

export const makeAugmentedSchema = (schema: GraphQLSchema, config?: AugmentConfig) => {
  const collectionMap = directiveTypeMap(schema, 'collection')
  const [pageMapSchemas, pageMapTypes] = buildPageMap({ collectionMap })
  const [filterMapSchemas, filterMapTypes] = buildFieldMap('filter', { collectionMap })
  const [insertMapSchemas, insertMapTypes] = buildFieldMap('insert', { collectionMap })
  const [updateMapSchemas, updateMapTypes] = buildFieldMap('update', { collectionMap })
  const [unsetMapSchemas, unsetMapTypes] = buildFieldMap('unset', { collectionMap })
  const [setMapSchemas, setMapTypes] = buildFieldMap('set', { collectionMap })
  const [incMapSchemas, incMapTypes] = buildFieldMap('inc', { collectionMap })
  const [decMapSchemas, decMapTypes] = buildFieldMap('dec', { collectionMap })
  // const resolvers = extractResolversFromSchema(schema)
  // const augmentedResolvers = augmentResolvers({
  //   resolvers,
  //   config
  // })
  return mapSchema(
    mergeSchemas({
      typeDefs: [],
      schemas: [
        schema,
        ...values(pageMapSchemas).filter(isSchema),
        ...values(filterMapSchemas).filter(isSchema),
        ...values(insertMapSchemas).filter(isSchema),
        ...values(updateMapSchemas).filter(isSchema),
        ...values(unsetMapSchemas).filter(isSchema),
        ...values(setMapSchemas).filter(isSchema),
        ...values(incMapSchemas).filter(isSchema),
        ...values(decMapSchemas).filter(isSchema),
      ],
    }),
    {
      [MapperKind.QUERY]: appendCrudQueryMapper({ collectionMap, filterMapTypes, pageMapTypes }),
      [MapperKind.MUTATION]: appendCrudMutationMapper({
        collectionMap,
        insertMapTypes,
        updateMapTypes,
        unsetMapTypes,
        setMapTypes,
        incMapTypes,
        decMapTypes,
      }),
    }
  )
  // return makeExecutableSchema({
  //   typeDefs: augmentedTypeDefs,
  //   resolvers: augmentedResolvers,
  //   // schemaDirectives: augmentedDirectives,
  //   resolverValidationOptions: {
  //     requireResolversForResolveType: false
  //   }
  // })
}

type CollectionMap = ReturnType<typeof directiveTypeMap>
function tuple<A, B>(a: A, b: B) {
  return [a, b] as [A, B]
}

const buildPageMap = ({ collectionMap }: { collectionMap: CollectionMap }) => {
  const types = mapValues(collectionMap, (collection, name) => {
    return new GraphQLObjectType({
      name: `${name}Page`,
      fields: {
        total: {
          type: GraphQLInt,
        },
        data: {
          type: new GraphQLList(collection),
        },
      },
    })
  })
  const schemas = mapValues(types, (type, name) => {
    return new GraphQLSchema({
      types: [type],
      query: new GraphQLObjectType({ name: 'Query', fields: {} }),
    })
  })
  return tuple(schemas, types)
}
const buildFieldMap = (type: string, { collectionMap }: { collectionMap: CollectionMap }) => {
  const schemas = mapValues(collectionMap, (collection, name) => {
    const fields = collection.astNode?.fields?.filter((field) => {
      return field.directives?.find((d) => d.name.value === type)
    })
    const types = fields
      ?.map((field) => `${field.name.value}: ${get('type.name.value', field)}`)
      .join('\n')
    if (!types?.length) return null
    return buildSchema(`input ${name}${capitalize(type)} {
      ${types}
    }`)
  })
  const types = mapValues(schemas, (schema, name) => {
    if (!schema) return null
    return schema.getType(`${name}${capitalize(type)}`) as GraphQLInputObjectType
  })
  return tuple(schemas, types)
}

const appendCrudQueryMapper = ({
  collectionMap,
  filterMapTypes,
  pageMapTypes,
}: {
  collectionMap: CollectionMap
  pageMapTypes: { [x: string]: GraphQLObjectType }
  filterMapTypes: { [x: string]: GraphQLInputObjectType | null }
}): ObjectTypeMapper => (type: GraphQLObjectType, schema: GraphQLSchema) => {
  const config = type.toConfig()
  const pagination = schema.getType('Pagination') as GraphQLInputObjectType
  const sort = schema.getType('Sort') as GraphQLInputObjectType
  const appendFields = Object.keys(collectionMap)
    .map((name) => {
      const collection = collectionMap[name]
      const find = {
        type: new GraphQLNonNull(pageMapTypes[name]),
        args: {
          pagination: {
            type: pagination,
          },
          sort: {
            type: sort,
          },
        } as any,
      }
      if (filterMapTypes[name]) {
        find.args.filter = {
          type: filterMapTypes[name],
        }
      }
      const findById = {
        type: collection,
        args: {
          id: {
            type: new GraphQLNonNull(graphqlTypeObjectId),
          },
        },
      }
      const findByIds = {
        type: new GraphQLList(collection),
        args: {
          ids: {
            type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(graphqlTypeObjectId))),
          },
        },
      }
      return {
        [`find${pluralize(name)}`]: find,
        [`find${name}ById`]: findById,
        [`find${pluralize(name)}ByIds`]: findByIds,
      }
    })
    .reduce((acc, fields) => ({ ...acc, ...fields }), {})
  return new GraphQLObjectType({
    ...config,
    fields: {
      ...config.fields,
      ...appendFields,
    },
  })
}

const appendCrudMutationMapper = ({
  collectionMap,
  insertMapTypes,
  updateMapTypes,
  unsetMapTypes,
  setMapTypes,
  incMapTypes,
  decMapTypes,
}: {
  collectionMap: CollectionMap
  insertMapTypes: { [x: string]: GraphQLInputObjectType | null }
  updateMapTypes: { [x: string]: GraphQLInputObjectType | null }
  unsetMapTypes: { [x: string]: GraphQLInputObjectType | null }
  setMapTypes: { [x: string]: GraphQLInputObjectType | null }
  incMapTypes: { [x: string]: GraphQLInputObjectType | null }
  decMapTypes: { [x: string]: GraphQLInputObjectType | null }
}): ObjectTypeMapper => (type: GraphQLObjectType, schema: GraphQLSchema) => {
  const config = type.toConfig()
  const appendFields = Object.keys(collectionMap)
    .map((name) => {
      const collection = collectionMap[name]
      const mutations: { [x: string]: any } = {}
      const insertType = insertMapTypes[name]
      const unsetType = unsetMapTypes[name]
      const setType = setMapTypes[name]
      const incType = incMapTypes[name]
      const decType = decMapTypes[name]
      if (insertType) {
        mutations[`insert${name}`] = {
          type: collection,
          args: {
            [camelCase(name)]: {
              type: new GraphQLNonNull(insertType),
            },
          },
        }
        mutations[`insertMany${pluralize(name)}`] = {
          type: new GraphQLList(collection),
          args: {
            [pluralize(camelCase(name))]: {
              type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(insertType))),
            },
          },
        }
      }
      if (unsetType || setType || incType || decType) {
        const updateArgs: { [x: string]: any } = {}
        if (unsetType) updateArgs[`${camelCase(name)}Unset`] = { type: unsetType }
        if (setType) updateArgs[`${camelCase(name)}Set`] = { type: setType }
        if (incType) updateArgs[`${camelCase(name)}Inc`] = { type: incType }
        if (decType) updateArgs[`${camelCase(name)}Dec`] = { type: decType }
        mutations[`update${name}`] = {
          type: collection,
          args: {
            id: {
              type: new GraphQLNonNull(graphqlTypeObjectId),
            },
            ...updateArgs,
          },
        }
        mutations[`updateMany${pluralize(name)}`] = {
          type: new GraphQLList(collection),
          args: {
            ids: {
              type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(graphqlTypeObjectId))),
            },
            ...updateArgs,
          },
        }
      }
      mutations[`remove${name}`] = {
        type: collection,
        args: {
          id: {
            type: new GraphQLNonNull(graphqlTypeObjectId),
          },
        },
      }
      mutations[`removeMany${pluralize(name)}`] = {
        type: new GraphQLList(collection),
        args: {
          ids: {
            type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(graphqlTypeObjectId))),
          },
        },
      }
      return mutations
    })
    .reduce((acc, fields) => ({ ...acc, ...fields }), {})
  return new GraphQLObjectType({
    ...config,
    fields: {
      ...config.fields,
      ...appendFields,
    },
  })
}

function transformMongoDirective(schema: GraphQLSchema, config: AugmentConfig) {
  const directives = schema.getDirectives()
  return directives.reduce((acc, directive) => {
    if (acc) acc[directive.name] = directive as any
    return acc
  }, {} as IExecutableSchemaDefinition['schemaDirectives'])
}

function extractDefinitions(schema: GraphQLSchema): DefinitionNode[] {
  const typeMap = schema.getTypeMap()
  return Object.keys(typeMap).reduce((acc, type) => {
    const ast = typeMap[type].astNode
    if (ast) {
      acc.push(ast)
      const extensionASTNodes = typeMap[type].extensionASTNodes
      if (extensionASTNodes) {
        acc.push(...extensionASTNodes)
      }
    }
    return acc
  }, [] as DefinitionNode[])
}

function pluralize(str: string) {
  if (last(str) === 's') return str
  return `${str}s`
}

export const extractResolversFromSchema = (schema: GraphQLSchema) => {
  const typeMap = schema.getTypeMap()
  const types = Object.keys(typeMap)
  let type = {}
  let schemaTypeResolvers = {} as { [key: string]: any } | undefined
  return types.reduce((acc, t) => {
    // prevent extraction from schema introspection system keys
    if (
      t !== '__Schema' &&
      t !== '__Type' &&
      t !== '__TypeKind' &&
      t !== '__Field' &&
      t !== '__InputValue' &&
      t !== '__EnumValue' &&
      t !== '__Directive'
    ) {
      type = typeMap[t]
      // resolvers are stored on the field level at a .resolve key
      schemaTypeResolvers = extractFieldResolversFromSchemaType(type)
      // do not add unless there exists at least one field resolver for type
      if (schemaTypeResolvers) {
        acc[t] = schemaTypeResolvers
      }
    }
    return acc
  }, {} as { [key: string]: { [key: string]: any } })
}

/**
 * Extracts field resolvers from a given type taken
 * from a schema
 */
const extractFieldResolversFromSchemaType = (type: any) => {
  const fields = type._fields
  const fieldKeys = fields ? Object.keys(fields) : []
  const fieldResolvers =
    fieldKeys.length > 0
      ? fieldKeys.reduce((acc, t) => {
          // do not add entry for this field unless it has resolver
          if (fields[t].resolve !== undefined) {
            acc[t] = fields[t].resolve
          }
          return acc
        }, {} as { [key: string]: any })
      : undefined
  // do not return value unless there exists at least 1 field resolver
  return fieldResolvers && Object.keys(fieldResolvers).length > 0 ? fieldResolvers : undefined
}
