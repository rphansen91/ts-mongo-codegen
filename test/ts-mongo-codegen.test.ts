import {
  plugin,
  addToSchema,
  mapFilterToMongo,
  mapUpdateToMongo,
  graphqlTypeObjectId,
  graphqlTypeDate
} from '../src/ts-mongo-codegen'
import { print, buildSchema, astFromValue, ValueNode, printSchema } from 'graphql'
import { ObjectID } from 'mongodb'
import { makeAugmentedSchema } from '../src/mongo-augment'

const bookSchema = buildSchema(`
${addToSchema.loc?.source?.body}

type Book @collection(name: "books") {
  id: ObjectId
  title: String @insert @set @unset @filter
  author: String @insert @set @unset @filter
}

type Query {
  root: String
}

type Mutation {
  root: String
}
`)
const schema = buildSchema(`
  ${addToSchema.loc?.source?.body}

  type User @collection(name: "user") {
    id: ObjectId
    email: String @insert @set @unset @filter
    username: String @insert @set @unset @filter
  }

  type Query {
    root: String
  }

  type Mutation {
    root: String
  }
`)

describe('TS Mongo Codegen', () => {
  it('Should convert to mongo filter', () => {
    expect(
      mapFilterToMongo({
        EQ: '$eq',
        GT: '$gt',
        GTE: '$gte',
        IN: ['1'],
        ALL: ['1', '2'],
        LT: '$lt',
        LTE: '$lte',
        NE: '$ne',
        NIN: ['1'],
        TEXT: {
          SEARCH: 'hello'
        }
      })
    ).toEqual({
      $eq: '$eq',
      $gt: '$gt',
      $gte: '$gte',
      $in: ['1'],
      $all: ['1', '2'],
      $lt: '$lt',
      $lte: '$lte',
      $ne: '$ne',
      $nin: ['1'],
      $text: {
        $search: 'hello'
      }
    })
  })

  it('Should convert to mongo update', () => {
    expect(
      mapUpdateToMongo({
        UNSET: {
          a: ''
        },
        SET: {
          a: 'b'
        },
        INC: { count: 1 },
        DEC: { count: 1 }
      })
    ).toEqual({
      $unset: {
        a: ''
      },
      $set: {
        a: 'b'
      },
      $inc: { count: 1 },
      $dec: { count: 1 }
    })
  })

  it('Should implement ObjectID Scalar', () => {
    const id = new ObjectID()
    const idSerialized = id.toHexString()
    const astValue = astFromValue(id, graphqlTypeObjectId) as ValueNode
    expect(graphqlTypeObjectId.parseLiteral(astValue, {})).toBeInstanceOf(ObjectID)
    expect(graphqlTypeObjectId.parseValue(idSerialized)).toBeInstanceOf(ObjectID)
    expect(graphqlTypeObjectId.serialize(id)).toBe(idSerialized)
  })

  it('Should implement Date Scalar', () => {
    const date = new Date()
    const dateSerialized = date.toISOString()
    const astValue = astFromValue(date, graphqlTypeDate) as ValueNode
    expect(graphqlTypeDate.parseLiteral(astValue, {})).toBeInstanceOf(Date)
    expect(graphqlTypeDate.parseValue(dateSerialized)).toBeInstanceOf(Date)
    expect(graphqlTypeDate.serialize(date)).toBe(dateSerialized)
  })
  it('Should parse date without ISOString', () => {
    const date = new Date()
    const dateSerialized = date.toString()
    expect(graphqlTypeDate.serialize({ toString: () => dateSerialized })).toBe(dateSerialized)
  })

  it('Should build crud operations for book', () => {
    const augmented = makeAugmentedSchema(bookSchema, {})
    const printed = printSchema(augmented)
    expect(printed).toMatchSnapshot('booksAugmentedSchema')
  })

  it('Should build crud operations', () => {
    const augmented = makeAugmentedSchema(schema, {})
    const printed = printSchema(augmented)
    expect(printed).toMatchSnapshot('usersAugmentedSchema')
  })

  it('Should build collection types', () => {
    expect(plugin(schema, [], {})).toMatchSnapshot('collectionTypes')
  })
})
