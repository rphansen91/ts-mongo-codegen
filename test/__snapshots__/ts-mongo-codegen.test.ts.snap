// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TS Mongo Codegen Should build collection types: collectionTypes 1`] = `
"import { Db, Collection } from 'mongodb'

export type IUserCollection = Collection<User>
export const getUserCollection = (db: Db) => db.collection<User>('user')

export function mongoCollectionFactory (db: Db) {
  const user = getUserCollection(db)

  return {
    user
  }
}"
`;

exports[`TS Mongo Codegen Should build crud operations for book: booksAugmentedSchema 1`] = `
"directive @collection(name: String!) on OBJECT

directive @filter on FIELD_DEFINITION

directive @insert on FIELD_DEFINITION

directive @update on FIELD_DEFINITION

directive @unset on FIELD_DEFINITION

directive @set on FIELD_DEFINITION

directive @inc on FIELD_DEFINITION

directive @dec on FIELD_DEFINITION

type Book {
  id: ObjectId
  title: String
  author: String
}

input BookFilter {
  title: String
  author: String
}

input BookInsert {
  title: String
  author: String
}

type BookPage {
  total: Int
  data: [Book]
}

input BookSet {
  title: String
  author: String
}

input BookUnset {
  title: String
  author: String
}

scalar Date

input DateFilter {
  EQ: Date
  GT: Date
  GTE: Date
  IN: [Date]
  ALL: [Date]
  LT: Date
  LTE: Date
  NE: Date
  NIN: [Date]
}

input FloatFilter {
  EQ: Float
  GT: Float
  GTE: Float
  IN: [Float]
  ALL: [Float]
  LT: Float
  LTE: Float
  NE: Float
  NIN: [Float]
}

input IntFilter {
  EQ: Int
  GT: Int
  GTE: Int
  IN: [Int]
  ALL: [Int]
  LT: Int
  LTE: Int
  NE: Int
  NIN: [Int]
}

type Mutation {
  root: String
  insertBook(book: BookInsert!): Book
  insertManyBooks(books: [BookInsert!]!): [Book]
  updateBook(id: ObjectId!, bookUnset: BookUnset, bookSet: BookSet): Book
  updateManyBooks(ids: [ObjectId!]!, bookUnset: BookUnset, bookSet: BookSet): [Book]
  removeBook(id: ObjectId!): Book
  removeManyBooks(ids: [ObjectId!]!): [Book]
}

scalar ObjectId

input ObjectIdFilter {
  EQ: ObjectId
  GT: ObjectId
  GTE: ObjectId
  IN: [ObjectId]
  ALL: [ObjectId]
  LT: ObjectId
  LTE: ObjectId
  NE: ObjectId
  NIN: [ObjectId]
}

input Pagination {
  perPage: Int
  page: Int
}

type Query {
  root: String
  findBooks(pagination: Pagination, sort: Sort, filter: BookFilter): BookPage!
  findBookById(id: ObjectId!): Book
  findBooksByIds(ids: [ObjectId!]!): [Book]
}

input Sort {
  field: String
  order: Int
}

input StringFilter {
  EQ: String
  GT: String
  GTE: String
  IN: [String]
  ALL: [String]
  LT: String
  LTE: String
  NE: String
  NIN: [String]
}
"
`;

exports[`TS Mongo Codegen Should build crud operations: usersAugmentedSchema 1`] = `
"directive @collection(name: String!) on OBJECT

directive @filter on FIELD_DEFINITION

directive @insert on FIELD_DEFINITION

directive @update on FIELD_DEFINITION

directive @unset on FIELD_DEFINITION

directive @set on FIELD_DEFINITION

directive @inc on FIELD_DEFINITION

directive @dec on FIELD_DEFINITION

scalar Date

input DateFilter {
  EQ: Date
  GT: Date
  GTE: Date
  IN: [Date]
  ALL: [Date]
  LT: Date
  LTE: Date
  NE: Date
  NIN: [Date]
}

input FloatFilter {
  EQ: Float
  GT: Float
  GTE: Float
  IN: [Float]
  ALL: [Float]
  LT: Float
  LTE: Float
  NE: Float
  NIN: [Float]
}

input IntFilter {
  EQ: Int
  GT: Int
  GTE: Int
  IN: [Int]
  ALL: [Int]
  LT: Int
  LTE: Int
  NE: Int
  NIN: [Int]
}

type Mutation {
  root: String
  insertUser(user: UserInsert!): User
  insertManyUsers(users: [UserInsert!]!): [User]
  updateUser(id: ObjectId!, userUnset: UserUnset, userSet: UserSet): User
  updateManyUsers(ids: [ObjectId!]!, userUnset: UserUnset, userSet: UserSet): [User]
  removeUser(id: ObjectId!): User
  removeManyUsers(ids: [ObjectId!]!): [User]
}

scalar ObjectId

input ObjectIdFilter {
  EQ: ObjectId
  GT: ObjectId
  GTE: ObjectId
  IN: [ObjectId]
  ALL: [ObjectId]
  LT: ObjectId
  LTE: ObjectId
  NE: ObjectId
  NIN: [ObjectId]
}

input Pagination {
  perPage: Int
  page: Int
}

type Query {
  root: String
  findUsers(pagination: Pagination, sort: Sort, filter: UserFilter): UserPage!
  findUserById(id: ObjectId!): User
  findUsersByIds(ids: [ObjectId!]!): [User]
}

input Sort {
  field: String
  order: Int
}

input StringFilter {
  EQ: String
  GT: String
  GTE: String
  IN: [String]
  ALL: [String]
  LT: String
  LTE: String
  NE: String
  NIN: [String]
}

type User {
  id: ObjectId
  email: String
  username: String
}

input UserFilter {
  email: String
  username: String
}

input UserInsert {
  email: String
  username: String
}

type UserPage {
  total: Int
  data: [User]
}

input UserSet {
  email: String
  username: String
}

input UserUnset {
  email: String
  username: String
}
"
`;
